stdlib/LinearAlgebra/src/symmetric.jl

```diff
@@ -414,6 +414,45 @@ function triu(A::Symmetric, k::Integer=0)
     end
 end
 
+function dot(A::Symmetric{Ta,<:AbstractArray}, B::Symmetric{Tb,<:AbstractArray}) where {Ta,Tb<:Number}
+    n = size(A, 2)
+    if n != size(B, 2)
+        throw(DimensionMismatch("A has dimensions $(size(A)) but B has dimensions $(size(B))"))
+    end
+
+    dotprod = zero(dot(first(A), first(B)))
+    @inbounds if A.uplo == 'U' && B.uplo == 'U'
+        for j in 1:n
+            for i in 1:(j - 1)
+                dotprod += 2 * dot(A.data[i, j], B.data[i, j])
+            end
+            dotprod += dot(A.data[j, j], B.data[j, j])
```

Comment:
To handle nested arrays, this needs to be 
```julia
dotprod += dot(symmetric(A.data[j, j], sym_uplo(A.uplo), symmetric(B.data[j, j], sym_uplo(B.uplo))
```
or maybe simply
```julia
dotprod += dot(A[j, j], B[j, j])
```
this will then take care of the symmetrization of diagonal elements.

And similarly below...